docs/design_proposals/filesystem-resolution-proposal-01.md

```diff
@@ -0,0 +1,110 @@
+# Filesystem Resolution 01
+
+* Author(s): cgwippern@google.com
+* Reviewers:
+* Date: 2020-02-12
+* Status: [Reviewed/Cancelled/Under implementation/Complete]
+
+Here is a brief explanation of the Statuses
+
+1. Reviewed: The proposal PR has been accepted, merged and ready for
+   implementation.
+2. Under implementation: An accepted proposal is being implemented by actual work.
+   Note: The design might change in this phase based on issues during
+   implementation.
+3. Cancelled: During or before implementation the proposal was cancelled.
+   It could be due to:
+   * other features added which made the current design proposal obsolete.
+   * No longer a priority.
+4. Complete: This feature/change is implemented.
+
+## Background
+
+Kaniko builds Docker image layers as overlay filesystem layers; specifically it
+creates a tar file which contains the entire content of a given layer in the
+overlay filesystem. Each overlay layer corresponds to one image layer.
+
+Overlay filesystems should only contain the objects changed in each layer;
+meaning that if only one file changes between some layer A and some B, layer B
+would only contain a single file (the one that changed).
+
+To accomplish this, Kaniko walks the entire filesystem to discover every object.
+Some of these objects may actually be a symlink to another object in the
+filesystem; in these cases we must consider both the link and the target object.
+
+Kaniko also maintains a set of whitelisted (aka ignored) filepaths. Any object
+which matches one of these filepaths should be ignored by kaniko.
+
+This results in a 3 dimensional search space
+
+* changed relative to previous layer
+* symlink
+* whitelisted
+
+Kaniko must also track which objects are referred to by multiple stages; this
+functionality is out of scope for this proposal.
+
+This search space is currently managed in an inconsistent and somewhat ad-hoc
+way; code that manages the various search dimensions is spread out and
+duplicated. There are also a number of poorly handled edge cases which continue
+to cause bugs.
+
+The search space dimensions cannot be reduced or substituted.
+
+Currently there are a number of bugs around symlinks incorrectly resolved,
+whitelists not respected, and unchanged files added to layers.
+
+## Design
+
+Resolution of the filesystem and the objects it contains should be handled
+consistently and with a single API.
+
+* Callers of this API should not be concerned with the type of object at a given filepath (e.g. symlink or not).
+* Callers of this API should not be concerned with whether a given path is whitelisted.
+* This API should return a set of filepaths which should be added to the layer
+  without further processing.
+
+The API should take a limited set of arguments
+* The root path in the filesystem
+* The whitelist
```

Comment:
I believe you are referring to this block of code
https://github.com/GoogleContainerTools/kaniko/blob/ae4e7d91aa6e825f6ac42d9dc085fda8d109cfd0/pkg/executor/build.go#L381

which as far as I can tell is currently only triggered by an uncached `RUN` command
https://github.com/GoogleContainerTools/kaniko/blob/ae4e7d91aa6e825f6ac42d9dc085fda8d109cfd0/pkg/commands/run.go#L137

So it seems the far more common case is passing a list of files.

Seems like we should actually separate this into two APIs. One that generates a list of files, one that filters/transforms the list of files.

Reply:
I don't know the full context on why some commands use the files api and why some scan from the root of the filesystem. I would guess it's an optimization for commands where we know for a fact which files were involved in the command. In the case of something like a RUN command its impossible to know which file were involved without scanning the filesystem.

This does present a challenge as the logic each command uses to generate the list of files it passes to the API is slightly different. Trying to unify, or handle any variation would be of great benefit.

Reply:
Any example of the comment above; in `CopyCommand` is `dest` is a symlink we only return the target in the set of files to snapshot; should we also be returning the link?

Reply:
> I would guess it's an optimization for commands where we know for a fact which files were involved in the command. 

+1. its for efficiency. There is no explanation in the design document (internal only) go/kaniko-design-doc.

Reply:
Lets for now keep this proposal simple. We are trying to tackle the problem which is an API for a given path, which will give you a list of files which should not need any further processing. 

We should look into `Copy` and `ADD` and see if parts of these functionalities could be reduced.

Reply:
>Lets for now keep this proposal simple. We are trying to tackle the problem which is an API for a given path, which will give you a list of files which should not need any further processing.

The API also needs to support a list of files as we mentioned. The generation of that list of files is also important. E.G what happens if a copy command only returns the targets of symlinks and not the link paths themselves?

Reply:
What are the dimensions the filtering logic need to handle:
* symlink/not-symlink
* whitelisted, not whitelisted

For the case of a symlink; what should be returned in the final output
There is the link, the target, and possibly the "real" link
IE
```
ln -s /foo/bar /baz/boom

touch /foo/bar/zoom.txt
```

link is `/baz/boom/zoom.txt`
target is `/foo/bar/zoom.txt`
the "real" link is `/baz/boom`

If my recent investigation is correct adding `/baz/boom/zoom.txt` to the layer tar will break the filesystem because the existing link `/bar/boom` makes `/bar/boom/zoom.txt` invalid.